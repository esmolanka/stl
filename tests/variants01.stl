
-- Empty variants unify
#check <> <: <>

-- Extra cases on the RHS unify with the polymoprphic tail on the LHS
#check <> <: <Foo : Unit>

-- Extra cases on the RHS _do_not_unify_ with absent fields on the LHS
#check <Foo : Unit> <: <>

-- Case order does not matter
#check <Foo : Unit | Bar : Void> <: <Bar : Void | Foo : Unit>

-- For extra cases, too
#check <Foo : Unit | Bar : Unit> <: <Baz : Unit | Bar : Unit | Foo : Unit>

-- Matching cases' types must unify
#check <Foo : Unit | Bar : Void> <: <Bar : Unit | Foo : Unit>
#check <Foo : Unit | Bar : {foo? : Unit}> <: <Bar : {bar? : Unit} | Foo : Unit>
