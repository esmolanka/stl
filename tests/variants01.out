tests/variants01.types:3:1-3:16:
  Subtype:
    ∀ (ρ : Row). < | … ρ>
  Supertype:
    ∀ (ρ : Row). < | … ρ>
  Result:
    OK
tests/variants01.types:6:1-6:26:
  Subtype:
    ∀ (ρ : Row). < | … ρ>
  Supertype:
    ∀ (ρ : Row). <Foo? : Unit | … ρ>
  Result:
    OK
tests/variants01.types:9:1-9:26:
  Subtype:
    ∀ (ρ : Row). <Foo? : Unit | … ρ>
  Supertype:
    ∀ (ρ : Row). < | … ρ>
  Result:
    FAIL
    Type is not polymorphic enough.
      Rigid variable ω does not unify with:
        ▫
tests/variants01.types:12:1-12:62:
  Subtype:
    ∀ (ρ : Row). <Foo? : Unit | Bar? : Void | … ρ>
  Supertype:
    ∀ (ρ : Row). <Bar? : Void | Foo? : Unit | … ρ>
  Result:
    OK
tests/variants01.types:15:1-15:75:
  Subtype:
    ∀ (ρ : Row). <Foo? : Unit | Bar? : Unit | … ρ>
  Supertype:
    ∀ (ρ : Row). <Baz? : Unit | Bar? : Unit | Foo? : Unit | … ρ>
  Result:
    OK
tests/variants01.types:18:1-18:62:
  Subtype:
    ∀ (ρ : Row). <Foo? : Unit | Bar? : Void | … ρ>
  Supertype:
    ∀ (ρ : Row). <Bar? : Unit | Foo? : Unit | … ρ>
  Result:
    FAIL
    Cannot unify types.
        Void
      Is not subsumed by:
        Unit
tests/variants01.types:19:1-19:80:
  Subtype:
    ∀ (ρ : Row). <Foo? : Unit | Bar? : ∃ (ρ : Row). {foo? : Unit, … ρ} | … ρ>
  Supertype:
    ∀ (ρ : Row). <Bar? : ∃ (ρ : Row). {bar? : Unit, … ρ} | Foo? : Unit | … ρ>
  Result:
    OK